refactor(core): allow React components for stage config

anotherchrisberry opened the pull request:
A pretty basic start to allowing us to write pipeline stage config as React components.

If a stage's config includes a `component` field, we'll use that. The component will get two props: `stage` and `stageFieldUpdated`, which is a dumb RxJS Subject that the stage should `next` whenever a value changes in a field (so the dirty flag on the pipeline is synced). Not loving that part, but not sure how better to handle it in this world.

jrsquared commented on app/scripts/modules/core/src/pipeline/config/stages/core/stageConfigField/StageConfigField.tsx:
```diff
@@ -0,0 +1,19 @@
+  <div className="form-group">
+    <label className="col-md-3 sm-label-right">
+      <span className="label-text">{props.label}</span>
+      {props.helpKey && <HelpField key={props.helpKey} />}
```
Is there a space between the label and the help icon?

jrsquared commented on app/scripts/modules/core/src/pipeline/config/stages/wait/WaitStageConfig.tsx:
```diff
@@ -0,0 +1,91 @@
+  }
+
+  public render() {
+    if (!this.state) {
```
since you're setting state in the constructor, will this ever be true?

anotherchrisberry commented:
Addressed questions from @jrsquared (thank you for catching those); also removed `Registry` from the Angular lifecycle and moved it to static methods. I'll split it out as a distinct export for `PipelineRegistry` in a separate PR...

christopherthielen commented on app/scripts/modules/core/src/pipeline/config/stages/stage.module.js:
```diff
@@ -147,7 +151,12 @@ module.exports = angular
       // clear existing contents
-      $element.find('.stage-details').html('');
+      if (reactComponentMounted) {
+        ReactDOM.unmountComponentAtNode(stageDetailsNode);
```
do we need to set `reactComponentMounted = false` here?

christopherthielen commented on app/scripts/modules/core/src/pipeline/config/stages/wait/WaitStageConfig.tsx:
```diff
@@ -0,0 +1,88 @@
+  private updateWaitTime = (waitTime: number | string) => {
+    this.props.stage.waitTime = waitTime;
+    this.setState({ waitTime });
+    this.props.stageFieldUpdated.next();
```
since no data is being provided to the subject, could we just use a callback?

anotherchrisberry replied:
yes!!!

christopherthielen commented on app/scripts/modules/core/src/pipeline/config/stages/wait/WaitStageConfig.tsx:
```diff
@@ -0,0 +1,88 @@
+    this.state = this.getState(props.stage);
+  }
+
+  private getState(stage: IStage): IWaitStageConfigState {
```
might be a good spot to try out `getDerivedStateFromProps()` https://reactjs.org/docs/react-component.html#static-getderivedstatefromprops

christopherthielen commented on app/scripts/modules/core/src/widgets/spelText/SpelNumberInput.tsx:
```diff
@@ -0,0 +1,87 @@
+export interface INumberInputState {
+  expressionActive: boolean;
+  glowing: boolean;
+}
```
you are glowing, chris
